migrations/2023-12-19-210053_tolerable-batch-insert-speed/up.sql

```diff
@@ -0,0 +1,166 @@
+-- Change triggers to run once per statement instead of once per row
+-- post_aggregates_post trigger doesn't need to handle deletion because the post_id column has ON DELETE CASCADE
+CREATE OR REPLACE FUNCTION post_aggregates_post ()
+    RETURNS TRIGGER
+    LANGUAGE plpgsql
+    AS $$
+BEGIN
+    INSERT INTO post_aggregates (post_id, published, newest_comment_time, newest_comment_time_necro, community_id, creator_id, instance_id)
+    SELECT
+        id,
+        published,
+        published,
+        published,
+        community_id,
+        creator_id,
+        (
+            SELECT
+                community.instance_id
+            FROM
+                community
+            WHERE
+                community.id = community_id
+            LIMIT 1)
+FROM
+    new_post;
+    RETURN NULL;
+END
+$$;
+
+CREATE OR REPLACE FUNCTION community_aggregates_post_count_insert ()
+    RETURNS TRIGGER
+    LANGUAGE plpgsql
+    AS $$
+BEGIN
+    UPDATE
+        community_aggregates
+    SET
+        posts = posts + post_group.count
+    FROM (
+        SELECT
+            community_id,
+            count(*)
+        FROM
+            new_post
+        GROUP BY
+            community_id) post_group
+WHERE
+    community_aggregates.community_id = post_group.community_id;
+    RETURN NULL;
+END
+$$;
+
+CREATE OR REPLACE FUNCTION person_aggregates_post_insert ()
+    RETURNS TRIGGER
+    LANGUAGE plpgsql
+    AS $$
+BEGIN
+    UPDATE
+        person_aggregates
+    SET
+        post_count = post_count + post_group.count
+    FROM (
+        SELECT
+            creator_id,
+            count(*)
+        FROM
+            new_post
+        GROUP BY
+            creator_id) post_group
+WHERE
+    person_aggregates.person_id = post_group.creator_id;
+    RETURN NULL;
+END
+$$;
+
+CREATE OR REPLACE TRIGGER post_aggregates_post
+    AFTER INSERT ON post REFERENCING NEW TABLE AS new_post
```

Comment:
From the postgres docs:

> The REFERENCING option enables collection of transition relations, which are row sets that include all of the rows inserted, deleted, or modified by the current SQL statement. This feature lets the trigger see a global view of what the statement did, not just one row at a time. This option is only allowed for an AFTER trigger that is not a constraint trigger; also, if the trigger is an UPDATE trigger, it must not specify a column_name list. OLD TABLE may only be specified once, and only for a trigger that can fire on UPDATE or DELETE; it creates a transition relation containing the before-images of all rows updated or deleted by the statement. Similarly, NEW TABLE may only be specified once, and only for a trigger that can fire on UPDATE or INSERT; it creates a transition relation containing the after-images of all rows updated or inserted by the statement.

Why is `REFERENCING` necessary? It should only be inserting one post_aggregates row anyway. 

I understand why it should be `for each statement`, but I'm not sure about these extras.

Reply:
Lemmy_db_perf (and possibly federation in the future) inserts multiple posts in 1 statement for better performance. There needs to be 1 post_aggregates row for each post row, so the optimization won't apply to the post_aggregates insertions unless the trigger is changed to do it with 1 statement, which is possible with `FOR EACH STATEMENT` but not `FOR EACH ROW`. Without the `REFERENCING` part, the trigger can't read the old or new post rows.

Inserting 1000000 posts (the amount recommended for query plan inspection [here](https://github.com/LemmyNet/lemmy/pull/4287#issuecomment-1861924509)) took almost 1 hour with the old triggers. It takes around 5 minutes with the new triggers.

Reply:
Nice, thx.